pyccel/ast/builtins.py

```diff
@@ -194,7 +194,7 @@ def __new__(cls, *args):
     def __init__(self, *args):
         if self.stage == 'syntactic' or len(args) == 0:
             return
-        is_homogeneous = all(args[0].dtype==a.dtype for a in args[1:])
+        is_homogeneous = all(False if a.dtype is NativeGeneric() else args[0].dtype==a.dtype for a in args[1:])
         is_homogeneous = is_homogeneous and all(args[0].rank==a.rank   for a in args[1:])
```

Comment:
It seems to me that these two lines could be made more readable by combining them into a single instruction:
```
is_homogeneous = all(a.dtype is not NativeGeneric() and args[0].dtype==a.dtype and args[0].rank==a.rank for a in args[1:])
```
Does that make sense?